From 4654eed7bbd833211d534da818c210c8b228ccba Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
Date: Sat, 16 Nov 2024 01:43:46 +0530
Subject: [PATCH] QSDK:CP:hostapd: add missing WDS repeater 6G bring-up, CSA
 implementation

Add frequency validation for new requested frequency during
interface start and add missing WDS repeater 6 GHz bring-up, CSA
implementation.

In WDS STA (5 GHz + 6 GHz) + repeater AP (2 GHz + 5 GHz + 6 GHz)
configured in MLO, some iterations of the STA connection result in
connection to only one link. During this scenario, the interface start
is called for all the disabled links of the AP with the associated
link frequency, causing two vdevs to be created in the same MLD
under the same radio. This commit adds frequency validation with the
channels information present in that radio, for the new requested
frequency during interface start.

Signed-off-by: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
---
 src/utils/ucode.c      | 95 ++++++++++++++++++++++++++++++++++--------
 wpa_supplicant/ucode.c | 18 +++++---
 2 files changed, 90 insertions(+), 23 deletions(-)

--- a/src/utils/ucode.c
+++ b/src/utils/ucode.c
@@ -110,10 +110,13 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	uc_value_t *freq = uc_fn_arg(0);
 	uc_value_t *sec = uc_fn_arg(1);
 	int width = ucv_uint64_get(uc_fn_arg(2));
+	uc_value_t *chan_width = uc_fn_arg(3);
+	int cf1 = ucv_uint64_get(uc_fn_arg(4));
+	int cf2 = ucv_uint64_get(uc_fn_arg(5));
 	int freq_val, center_idx, center_ofs;
 	enum oper_chan_width chanwidth;
 	enum hostapd_hw_mode hw_mode;
-	u8 op_class, channel, tmp_channel;
+	u8 op_class, channel, tmp_channel, seg0_idx = 0, seg1_idx = 0;
 	const char *modestr;
 	int sec_channel = 0;
 	uc_value_t *ret;
@@ -135,19 +138,47 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 		return NULL;
 
 	switch (width) {
-	case 0:
-		chanwidth = CONF_OPER_CHWIDTH_USE_HT;
-		break;
-	case 1:
-		chanwidth = CONF_OPER_CHWIDTH_80MHZ;
-		break;
-	case 2:
-		chanwidth = CONF_OPER_CHWIDTH_160MHZ;
-		break;
-	default:
-		return NULL;
-	}
+       case 0:
+               chanwidth = CONF_OPER_CHWIDTH_USE_HT;
+               break;
+       case 1:
+               chanwidth = CONF_OPER_CHWIDTH_80MHZ;
+               break;
+       case 2:
+               chanwidth = CONF_OPER_CHWIDTH_160MHZ;
+               break;
+       case 9:
+               chanwidth = CONF_OPER_CHWIDTH_320MHZ;
+               break;
+       default:
+               return NULL;
+       }
+
+	if (ucv_type(chan_width) != UC_INTEGER)
+		goto skip_chan_width;
+
+	switch (ucv_int64_get(chan_width)) {
+        case CHAN_WIDTH_80:
+                chanwidth = CONF_OPER_CHWIDTH_80MHZ;
+                break;
+        case CHAN_WIDTH_80P80:
+                chanwidth = CONF_OPER_CHWIDTH_80P80MHZ;
+                break;
+        case CHAN_WIDTH_160:
+                chanwidth = CONF_OPER_CHWIDTH_160MHZ;
+                break;
+        case CHAN_WIDTH_320:
+                chanwidth = CONF_OPER_CHWIDTH_320MHZ;
+                break;
+        case CHAN_WIDTH_20_NOHT:
+        case CHAN_WIDTH_20:
+        case CHAN_WIDTH_40:
+        default:
+                chanwidth = CONF_OPER_CHWIDTH_USE_HT;
+                break;
+        }
 
+skip_chan_width:
 	hw_mode = ieee80211_freq_to_channel_ext(freq_val, sec_channel,
 						chanwidth, &op_class, &channel);
 	switch (hw_mode) {
@@ -174,8 +205,9 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	ucv_object_add(ret, "hw_mode_str", ucv_get(ucv_string_new(modestr)));
 	ucv_object_add(ret, "sec_channel", ucv_int64_new(sec_channel));
 	ucv_object_add(ret, "frequency", ucv_int64_new(freq_val));
+	ucv_object_add(ret, "oper_chwidth", ucv_int64_new(chanwidth));
 
-	if (!sec_channel)
+	if (!sec_channel && freq_val < 5900)
 		return ret;
 
 	if (freq_val >= 5900)
@@ -188,6 +220,10 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	tmp_channel &= ~((8 << width) - 1);
 	center_idx = tmp_channel + center_ofs + (4 << width) - 1;
 
+	// Handle EHT240 in 5G
+	if (freq_val >= 5500 && freq_val <= 5730 && chanwidth == 9)
+		center_idx=130;
+
 	if (freq_val < 3000)
 		ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(0));
 	else
@@ -195,8 +231,37 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	center_idx = (center_idx - channel) * 5 + freq_val;
 	ucv_object_add(ret, "center_freq1", ucv_int64_new(center_idx));
 
-	wpa_printf(MSG_INFO, "[debug] %s: op_c: %d, ch:%d, hw_mode:%d, mode:%s, sec_ch:%d\n freq:%d, center_seg0:%d \n",
-		__func__, op_class, channel, hw_mode, modestr, sec_channel, freq_val, center_idx);
+	if (!cf1 && !cf2)
+		return ret;
+
+	switch (hw_mode) {
+        case HOSTAPD_MODE_IEEE80211A:
+                if (cf1 == 5935)
+                        seg0_idx = (cf1 - 5925) / 5;
+                else if (cf1 > 5950)
+                        seg0_idx = (cf1 - 5950) / 5;
+                else if (cf1 > 5000)
+                        seg0_idx = (cf1 - 5000) / 5;
+
+                if (cf2 == 5935)
+                        seg1_idx = (cf2 - 5925) / 5;
+                else if (cf2 > 5950)
+                        seg1_idx = (cf2 - 5950) / 5;
+                else if (cf2 > 5000)
+                        seg1_idx = (cf2 - 5000) / 5;
+                break;
+        default:
+		if (cf1)
+			ieee80211_freq_to_chan(cf1, &seg0_idx);
+		if (cf2)
+			ieee80211_freq_to_chan(cf2, &seg1_idx);
+                break;
+        }
+
+	ucv_object_add(ret, "center_freq1", ucv_int64_new(cf1));
+	ucv_object_add(ret, "center_freq2", ucv_int64_new(cf2));
+	ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(seg0_idx));
+	ucv_object_add(ret, "center_seg1_idx", ucv_int64_new(seg1_idx));
 
 out:
 	return ret;
--- a/wpa_supplicant/ucode.c
+++ b/wpa_supplicant/ucode.c
@@ -158,6 +158,7 @@ void wpas_ucode_event(struct wpa_supplic
 	if (event == EVENT_CH_SWITCH_STARTED) {
 		ucv_object_add(val, "csa_count", ucv_int64_new(data->ch_switch.count));
 		ucv_object_add(val, "frequency", ucv_int64_new(data->ch_switch.freq));
+		ucv_object_add(val, "chan_width", ucv_int64_new(data->ch_switch.ch_width));
 		ucv_object_add(val, "sec_chan_offset", ucv_int64_new(data->ch_switch.ch_offset));
 		ucv_object_add(val, "center_freq1", ucv_int64_new(data->ch_switch.cf1));
 		ucv_object_add(val, "center_freq2", ucv_int64_new(data->ch_switch.cf2));
@@ -281,6 +282,7 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 	uc_value_t *radio_id = uc_fn_arg(0);
 	s8 hw_idx;
 	int freq, i;
+	struct wpa_signal_info si = {0};
 	struct wpa_mlo_signal_info mlo_si = {0};
 
 	if (!wpa_s)
@@ -294,10 +296,8 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 	hw_idx = ucv_int64_get(radio_id);
 
 	if (wpa_s->wpa_state == WPA_COMPLETED && wpa_s->valid_links) {
-		wpa_printf(MSG_INFO, "%s: wpa_s->valid_links=%d and hw_idx %d", __func__, wpa_s->valid_links, hw_idx);
 		for_each_link(wpa_s->valid_links, i) {
 			freq = wpa_s->links[i].freq;
-			wpa_printf(MSG_INFO, "%s: freq: %d for link_id %d", __func__, freq, i);
 			if ( hw_idx == wpa_get_hw_idx_by_freq(wpa_s, freq )) {
 				ucv_object_add(ret, "frequency", ucv_int64_new(freq));
 				if (wpa_drv_mlo_signal_poll(wpa_s, &mlo_si) == 0) {
@@ -305,11 +305,7 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 						ucv_object_add(ret, "chan_width", ucv_int64_new(mlo_si.links[i].chanwidth));
 						ucv_object_add(ret, "center_freq1", ucv_int64_new(mlo_si.links[i].center_frq1));
 						ucv_object_add(ret, "center_freq2", ucv_int64_new(mlo_si.links[i].center_frq2));
-						wpa_printf(MSG_INFO, "%s: status frequency filled %d width %d cf1 %d cf2 %d for radio_id %d", __func__, freq,
-						mlo_si.links[i].chanwidth, mlo_si.links[i].center_frq1, mlo_si.links[i].center_frq2, hw_idx);
 					}
-				} else {
-						wpa_printf(MSG_ERROR, "%s: error getting signal poll");
 				}
 			}
 		}
@@ -336,6 +332,17 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 
 		ucv_object_add(ret, "sec_chan_offset", ucv_int64_new(sec_chan));
 		ucv_object_add(ret, "frequency", ucv_int64_new(bss->freq));
+		if (wpa_s->wpa_state == WPA_COMPLETED &&
+		    wpa_drv_signal_poll(wpa_s, &si) == 0) {
+			if (si.chanwidth != CHAN_WIDTH_UNKNOWN) {
+				ucv_object_add(ret, "chan_width",
+					       ucv_int64_new(si.chanwidth));
+				ucv_object_add(ret, "center_freq1",
+					       ucv_int64_new(si.center_frq1));
+				ucv_object_add(ret, "center_freq2",
+					       ucv_int64_new(si.center_frq2));
+			}
+		}
 	}
 
 #ifdef CONFIG_MESH
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -2473,7 +2473,7 @@ static int hostapd_ctrl_register_frame(s
 
 #ifdef NEED_AP_MLME
 
-static bool
+bool
 hostapd_ctrl_is_freq_in_cmode(struct hostapd_hw_modes *mode,
 			      struct hostapd_multi_hw_info *current_hw_info,
 			      int freq)
--- a/src/ap/ctrl_iface_ap.h
+++ b/src/ap/ctrl_iface_ap.h
@@ -62,5 +62,8 @@ int hostapd_ctrl_iface_acl_show_mac(stru
 				    char *buf, size_t buflen);
 int hostapd_disassoc_accept_mac(struct hostapd_data *hapd);
 int hostapd_disassoc_deny_mac(struct hostapd_data *hapd);
+bool hostapd_ctrl_is_freq_in_cmode(struct hostapd_hw_modes *mode,
+				   struct hostapd_multi_hw_info *current_hw_info,
+				   int freq);
 u8 hostapd_maxnss(struct hostapd_data *hapd, struct sta_info *sta);
 #endif /* CTRL_IFACE_AP_H */
--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -11,6 +11,7 @@
 #include "acs.h"
 #include <libubox/uloop.h>
 #include "sta_info.h"
+#include "ctrl_iface_ap.h"
 
 static uc_resource_type_t *global_type, *bss_type, *iface_type;
 static struct hapd_interfaces *interfaces;
@@ -578,7 +579,7 @@ uc_hostapd_iface_start(uc_vm_t *vm, size
 	struct hostapd_config *conf;
 	bool changed = false;
 	uint64_t intval;
-	int i;
+	int i, freq;
 
 	if (!iface)
 		return NULL;
@@ -608,6 +609,22 @@ uc_hostapd_iface_start(uc_vm_t *vm, size
 	     iface->state == HAPD_IFACE_DFS))
 		return ucv_boolean_new(true);
 
+	intval = ucv_int64_get(ucv_object_get(info, "frequency", NULL));
+	if (!errno) {
+		freq = (int)intval;
+		if (!hostapd_ctrl_is_freq_in_cmode(iface->current_mode,
+						   iface->current_hw_info,
+						   freq)) {
+			wpa_printf(MSG_ERROR, "Interface start requested with freq=%d but previous iface freq=%d\n",
+				   freq, iface->freq);
+			return NULL;
+		}
+	} else {
+		freq = 0;
+	}
+
+	iface->freq = freq;
+
 	intval = ucv_int64_get(ucv_object_get(info, "center_seg0_idx", NULL));
 	if (!errno)
 		hostapd_set_oper_centr_freq_seg0_idx(conf, intval);
@@ -620,11 +637,6 @@ uc_hostapd_iface_start(uc_vm_t *vm, size
 	if (!errno)
 		hostapd_set_oper_chwidth(conf, intval);
 
-	intval = ucv_int64_get(ucv_object_get(info, "frequency", NULL));
-	if (!errno)
-		iface->freq = intval;
-	else
-		iface->freq = 0;
 	conf->acs = 0;
 
 out:
